### DISCUSSION TITLE
validate if a value is in a list

### DISCUSSION THREAD

**Discussion by ahmed711** [cue-lang/cue#1835] [Q&A] [answered]
Hi all, 
I am new to cue and I am still wrapping my head around different options, so excuse my ignorance 

In this example we have 

templates: [...template]
#template: {
  name: 
  length:
  width:
}

drawings:[...drawing]
drawing: {
  usedTemplate: string & in #template.name
  ...
  } 

I want to validate that usedTemplate in drawing already exists in templates, I would expect it to be simple like <if value in a list> but I am not sure to be honest

**Comment by slewiskelly** [comment, marked as answer]
The example provided was incomplete so I went ahead and added a couple of changes so there is a valid starting point:

```cue
templates: [...template]

#template: {
	name:   string
	length: int
	width:  int
}

drawings: [...drawing]

drawing: {
	usedTemplate: string // & in #template.name
	...
}

```
-----

There are a couple of ways, which one you decide is up to your specific requirements.

1. You can use a combinarion of [list comprehension](https://cuelang.org/docs/tutorials/tour/expressions/listcomp/) and the [`or` builtin](https://cuelang.org/docs/references/spec/#or)

```cue
#drawing: {
	usedTemplate: or([ for t in templates {t.name}])
	...
}

drawings: [...#drawing]

#template: {
	name:   string
	length: int
	width:  int
}

templates: [...#template]

templates: [
    {name: "small", length: 1, width: 1},
    {name: "medium", length: 2, width: 3},
    {name: "large", length: 3, width: 3},
]

drawings: [
    {usedTemplate: "small"},
    {usedTemplate: "medium"},
    {usedTemplate: "large"},
    // {usedTemplate: "extra-large"},
]
```

2. If you can, you can also use a [disjunction](https://cuelang.org/docs/tutorials/tour/types/disjunctions/) to constrain the template name as well as the template of the drawing.

```cue
#drawing: {
	usedTemplate: #templateType
	...
}

drawings: [...#drawing]

#template: {
	name:   #templateType
	length: int
	width:  int
}

#templateType: "small" | "medium" | "large"

templates: [...#template]

templates: [
    {name: "small", length: 1, width: 1},
    {name: "medium", length: 2, width: 3},
    {name: "large", length: 3, width: 3},
]

drawings: [
    {usedTemplate: "small"},
    {usedTemplate: "medium"},
    {usedTemplate: "large"},
    // {usedTemplate: "extra-large"},
]
```

Nevertheless, in both cases, you'll receive an error if you uncomment the last drawing.
